### PR TITLE
Add Bullet k-DOP and occlusion culling test

### PR DESCRIPTION
**Pull request by jsjolund**
Occlusion culling is a camera frustum culling technique which calculates not only whether an object is inside the frustum, but also if it is visible from the point of view of the camera.

k-DOP culling (k-Discrete Oriented Polytope) is a more conventional frustum culling, which uses the broadphase bounding volume tree to accelerate the culling. In this way, not all objects in the world need to be checked whether or not they are inside the frustum.

Demo: http://youtu.be/c1ncEbdDRvo

(Resubmission of #3730, I could not figure out how to squash the commits correctly when branches diverged...)

### CODE REVIEW

**Comment by jsjolund** [comment]
I finally got around to testing this on an Android phone, a Motorola Nexus 6, but it crashed. Not sure what the problem is?
```
01-19 05:45:34.032: I/TestCollection(9716): created test 'com.badlogic.gdx.tests.bullet.OcclusionCullingTest'
01-19 05:45:34.042: A/libc(9716): Fatal signal 11 (SIGSEGV), code 1, fault addr 0xdeadcab1 in tid 9839 (GLThread 2710)
01-19 05:45:34.096: A/DEBUG(1720): *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
01-19 05:45:34.096: A/DEBUG(1720): Build fingerprint: 'google/shamu/shamu:6.0.1/MMB29S/2489379:user/release-keys'
01-19 05:45:34.096: A/DEBUG(1720): Revision: '0'
01-19 05:45:34.097: A/DEBUG(1720): ABI: 'arm'
01-19 05:45:34.098: A/DEBUG(1720): pid: 9716, tid: 9839, name: GLThread 2710  >>> com.badlogic.gdx.tests.android <<<
01-19 05:45:34.098: A/DEBUG(1720): signal 11 (SIGSEGV), code 1 (SEGV_MAPERR), fault addr 0xdeadcab1
01-19 05:45:34.131: A/DEBUG(1720):     r0 9e87e325  r1 9e87e325  r2 00000000  r3 deadcab1
01-19 05:45:34.131: A/DEBUG(1720):     r4 981b2800  r5 981b2838  r6 b3944000  r7 474e5543
01-19 05:45:34.131: A/DEBUG(1720):     r8 74fc1c28  r9 9b9852c0  sl 981cb490  fp 97cbe7c0
01-19 05:45:34.131: A/DEBUG(1720):     ip 9e8b5fc8  sp 9b9850f0  lr 9e87e33f  pc 9e87e330  cpsr 600e0030
01-19 05:45:34.160: A/DEBUG(1720): backtrace:
01-19 05:45:34.160: A/DEBUG(1720):     #00 pc 002f3330  /data/app/com.badlogic.gdx.tests.android-1/lib/arm/libgdx-bullet.so
01-19 05:45:34.160: A/DEBUG(1720):     #01 pc 002f333d  /data/app/com.badlogic.gdx.tests.android-1/lib/arm/libgdx-bullet.so
01-19 05:45:34.161: A/DEBUG(1720):     #02 pc 002f33bf  /data/app/com.badlogic.gdx.tests.android-1/lib/arm/libgdx-bullet.so
01-19 05:45:34.161: A/DEBUG(1720):     #03 pc 002f2a5f  /data/app/com.badlogic.gdx.tests.android-1/lib/arm/libgdx-bullet.so
01-19 05:45:34.161: A/DEBUG(1720):     #04 pc 002f23fd  /data/app/com.badlogic.gdx.tests.android-1/lib/arm/libgdx-bullet.so
01-19 05:45:34.161: A/DEBUG(1720):     #05 pc 002f24c3  /data/app/com.badlogic.gdx.tests.android-1/lib/arm/libgdx-bullet.so
01-19 05:45:34.161: A/DEBUG(1720):     #06 pc 0023a3b8  /data/app/com.badlogic.gdx.tests.android-1/lib/arm/libgdx-bullet.so (SwigDirector_ICollide::Process(btDbvtNode const*, float)+488)
01-19 05:45:34.161: A/DEBUG(1720):     #07 pc 00289c44  /data/app/com.badlogic.gdx.tests.android-1/lib/arm/libgdx-bullet.so (btDbvt::collideOCL(btDbvtNode const*, btVector3 const*, float const*, btVector3 const&, int, btDbvt::ICollide&, bool)+2160)
01-19 05:45:34.161: A/DEBUG(1720):     #08 pc 0028a658  /data/app/com.badlogic.gdx.tests.android-1/lib/arm/libgdx-bullet.so (Java_com_badlogic_gdx_physics_bullet_collision_CollisionJNI_btDbvt_1collideOCL_1_1SWIG_13+180)
01-19 05:45:34.161: A/DEBUG(1720):     #09 pc 009a0f01  /data/app/com.badlogic.gdx.tests.android-1/oat/arm/base.odex (offset 0x766000) (void com.badlogic.gdx.physics.bullet.collision.CollisionJNI.btDbvt_collideOCL__SWIG_3(long, com.badlogic.gdx.physics.bullet.collision.btDbvtNode, java.nio.FloatBuffer, java.nio.FloatBuffer, com.badlogic.gdx.math.Vector3, int, long, com.badlogic.gdx.physics.bullet.collision.ICollide)+244)
01-19 05:45:34.161: A/DEBUG(1720):     #10 pc 00fa5c29  /data/app/com.badlogic.gdx.tests.android-1/oat/arm/base.odex (offset 0x766000) (void com.badlogic.gdx.physics.bullet.collision.btDbvt.collideOCL(com.badlogic.gdx.physics.bullet.collision.btDbvtNode, java.nio.FloatBuffer, java.nio.FloatBuffer, com.badlogic.gdx.math.Vector3, int, com.badlogic.gdx.physics.bullet.collision.ICollide)+644)
01-19 05:45:34.161: A/DEBUG(1720):     #11 pc 00dd2c83  /data/app/com.badlogic.gdx.tests.android-1/oat/arm/base.odex (offset 0x766000) (void com.badlogic.gdx.tests.bullet.OcclusionCuller.performOcclusionCulling(com.badlogic.gdx.physics.bullet.collision.btDbvtBroadphase, com.badlogic.gdx.tests.bullet.OcclusionBuffer, com.badlogic.gdx.graphics.Camera)+502)
01-19 05:45:34.162: A/DEBUG(1720):     #12 pc 012f303b  /data/app/com.badlogic.gdx.tests.android-1/oat/arm/base.odex (offset 0x766000) (void com.badlogic.gdx.tests.bullet.OcclusionCullingTest.renderWorld()+1222)
01-19 05:45:34.162: A/DEBUG(1720):     #13 pc 010fc79f  /data/app/com.badlogic.gdx.tests.android-1/oat/arm/base.odex (offset 0x766000) (void com.badlogic.gdx.tests.bullet.BaseBulletTest.render(boolean)+250)
01-19 05:45:34.162: A/DEBUG(1720):     #14 pc 010fc677  /data/app/com.badlogic.gdx.tests.android-1/oat/arm/base.odex (offset 0x766000) (void com.badlogic.gdx.tests.bullet.BaseBulletTest.render()+58)
01-19 05:45:34.162: A/DEBUG(1720):     #15 pc 012f2729  /data/app/com.badlogic.gdx.tests.android-1/oat/arm/base.odex (offset 0x766000) (void com.badlogic.gdx.tests.bullet.OcclusionCullingTest.render()+60)
01-19 05:45:34.162: A/DEBUG(1720):     #16 pc 0123cbbd  /data/app/com.badlogic.gdx.tests.android-1/oat/arm/base.odex (offset 0x766000) (void com.badlogic.gdx.tests.BulletTestCollection.render()+216)
01-19 05:45:34.162: A/DEBUG(1720):     #17 pc 00c97fff  /data/app/com.badlogic.gdx.tests.android-1/oat/arm/base.odex (offset 0x766000) (void com.badlogic.gdx.backends.android.AndroidGraphics.onDrawFrame(javax.microedition.khronos.opengles.GL10)+2074)
01-19 05:45:34.162: A/DEBUG(1720):     #18 pc 744c2099  /data/dalvik-cache/arm/system@framework@boot.oat (offset 0x1ec9000)

```

**Comment by xoppa** [comment]
most commonly you'll notice problems on android sooner because the gc kicks in earlier. Try adding `System.gc();` around the place where it crashes and test whether it crashes on desktop as well.

**Comment by jsjolund** [comment]
I traced the problem to the [getCollisionObject](https://github.com/jsjolund/libgdx/blob/master/tests/gdx-tests/src/com/badlogic/gdx/tests/bullet/OcclusionCuller.java#L96) method. It works fine on desktop, but on Android, btDbvtNode#getData() returns a negative number, which I don't think is valid (but not sure). This never happens on desktop. The [data field](http://bulletphysics.org/Bullet/BulletFull/btDbvt_8h_source.html#l00186) is defined as a void pointer, and is supposed to point to a btBroadphaseProxy instance. This is what I got, running it through the debugger in Idea:

**Comment by xoppa** [comment]
> The data field is defined as a void pointer, and is supposed to point to a btBroadphaseProxy instance.
Where does it state that it is supposed to point to a `btBroadphaseProxy`? Since [it's a union](https://github.com/libgdx/libgdx/blob/master/extensions/gdx-bullet/jni/src/bullet/BulletCollision/BroadphaseCollision/btDbvt.h#L183-L188), I would expect it to point to a `btDbvtNode` if anything at all. Could it be that it works for you on desktop because that is 64 bit? Perhaps try 32 bit desktop if you haven't done so already.

**Comment by jsjolund** [comment]
It is cast to a btBroadphaseProxy in [CDTestFramework/BulletSAPCompleteBoxPruningTest.cpp](https://github.com/kripken/bullet/blob/master/Extras/CDTestFramework/BulletSAPCompleteBoxPruningTest.cpp#L894).

**Comment by jsjolund** [comment]
I will look into the architectures as the source of the problem, thanks.

**Comment by jsjolund** [comment]
Actually, it is not ```btDbvtNode#getData()``` which is the problem, this value seems to map to a valid ```btBroadphaseProxy``` when running:
```btBroadphaseProxy proxy = btBroadphaseProxy.internalTemp(node.getData() false);```

The problem actually occurs on:
```btCollisionObject object = btCollisionObject.getInstance(proxy.getClientObject());```
This does not seem to yield a valid ```btCollisionObject```, but null (on Android, it works on desktop). I will have to look into how the pooling in this class works with respect to swigCPtr and such.

File: extensions/gdx-bullet/jni/swig/collision/btDbvt.i [hunk 1]
```diff
@@ -102,5 +102,13 @@
 	btDbvtNode *getChild1() {
 		return $self->childs[1];
 	}
+	
+	btBroadphaseProxy *getDataAsProxy() {
+	    return (btBroadphaseProxy*)$self->data;
+	}
+	
+	btCollisionObject *getDataAsProxyClientObject() {
```
Comment (jsjolund) [inline extensions/gdx-bullet/jni/swig/collision/btDbvt.i]: As far as I understand it, every ```btDbvtNode``` points to a ```btBroadphaseProxy```, but not all ```btBroadphaseProxy``` points to a ```btCollisionObject```.

File: extensions/gdx-bullet/jni/swig/collision/collision.i [hunk 2]
```diff
@@ -299,6 +295,8 @@ void btMultiSapBroadphase::quicksort(btBroadphasePairArray& a, int lo, int hi)
 %include "BulletCollision/CollisionShapes/btConvex2dShape.h"
 
 %include "./btCollisionObject.i"
+%include "./btDbvt.i"
```
Comment (jsjolund) [inline extensions/gdx-bullet/jni/swig/collision/collision.i]: Needed to change orders of includes to be able to add the btDbvtNode getters.

**Comment by jsjolund** [comment]
Works on Android now, performance increase for this occlusion culling is less than on desktop, probably because of weaker CPU. Still, on my Motorola Nexus 6 phone, there was around 10 FPS improvement when around 1000 models were in the world.

**Comment by jsjolund** [comment]
As a side note, this type of culling becomes more worthwhile if you have a lot of high detail models, textures, shader effects and similar. If they can be "hidden" behind walls, and therefore not have to be rendered, at some point the cost of drawing them with GPU will be higher than the cost of performing an occlusion query with CPU.

Another thing to note is that this does not play well with ```DirectionalShadowLight```. As can be seen in the demo, shadows tend to flicker when an object is close to the edge of an occluding wall.

**Comment by jsjolund** [comment]
Should I make a separate pull request for only the Bullet changes again?

**Comment by xoppa** [comment]
Thanks!

**Pull request merged**